### PR TITLE
Improve typing for inferred project

### PR DESCRIPTION
**Pull request by billti**
Porting #13641 and bringing it up to date so that pull request can be closed, as it was old and unmergable, and Zhengbo is no longer on the team. I've done little testing or in depth reviewing yet. 

Details from that pull request:

Fixes a VSO issue.

### Background:
Currently for an external project that contains tsconfig files, we put files included by the tsconfig file in a configured projects, and the rest of files in the miscellaneous context. This behavior caused problems for type acquisition, because:
1. the project specific types are not used for these misc files, for example cordova and sharepoint typing files;
2. these misc files do not have a "project root" for type discovery purpose. 

Two fix these two issues, I have made the following changes in this PR:
1. Added an API called "setTypeAcquisitionForInferredProject", which can add all the project specific types to the inferred project;
2. Added some logic so that during type discovery, if no "project root path" is provided, try to go up and guess the first possible project root path.

### CODE REVIEW

File: src/server/editorServices.ts [hunk 1]
```diff
@@ -1301,6 +1310,8 @@ namespace ts.server {
                 ? this.inferredProjects[0]
                 : new InferredProject(this, this.documentRegistry, this.compilerOptionsForInferredProjects);
 
+            project.setTypeAcquisition(this.typeAcquisitionForInferredProjects);
```
Comment (sheetalkamat) [inline src/server/editorServices.ts]: Can you please pass ```this.typeAcquisitionForInferredProjects``` to the constructor (like ```this.compilerOptionsForInferredProjects```) instead

**Comment by mhegazy** [comment]
@billti is this still needed?

**Comment by billti** [comment]
I still have it on my TODO list to revisit this end ensure the bugs/behavior it was intended to address, have been resolved already by other changes. I know at least some have already been fixed (by changes from @sheetalkamat I believe).

**Comment by billti** [comment]
And not to be the bearer of bad news, but it is currently number `#42` on my todo list 😱 . I can bump it up or reassign if you need this addressed in the near term.

**Comment by RyanCavanaugh** [comment]
Closing due to inactivity. Seems like recent work in this area has addressed most of the original issues.

**Pull request closed without merging**